File: LeetCode_Algorithm/51(N-Queens)/51.py
from turtle import back
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        # sol1 171ms 29%
        self.temp = [['.'] * n for _ in range(n)]
        self.ans = []
        self.dfs(0)
        return self.ans
    def dfs(self, r):
        if r == len(self.temp):
            temp = []
            for r in range(len(self.temp)):
                temp.append("".join(self.temp[r]))
            self.ans.append(temp); 
            return ;
        for c in range(len(self.temp[0])):
            if self.temp[r][c] != 'Q':
                self.temp[r][c] = 'Q'
                if self.Check(r-1, c):
                    self.dfs(r+1)
                self.temp[r][c] = '.'
    def Check(self, r,c):
        backslash = c-1; slash = c+1
        while -1 < r:
            if -1 < backslash and self.temp[r][backslash] == 'Q':
                return False
            if slash < len(self.temp[0]) and self.temp[r][slash] == 'Q':
                return False
            if self.temp[r][c] == 'Q':
                return False
            r -= 1; slash += 1; backslash -= 1
        return True
                
print(Solution().solveNQueens(4))
print(Solution().solveNQueens(5))

File: LeetCode_Algorithm/51(N-Queens)/51.js
var solveNQueens = function(n) {
  // sol1 126ms 50%
  ans = []
  board = Array.from(new Array(n), () =>new Array(n).fill('.'))
  dfs(0)
  return ans
};
var dfs = function(r) {
  if (r == board.length) {
    let temp = []
    for (let i=0; i < board.length; i++) {
      temp.push(board[i].join(''))
    }
    ans.push(temp)
    return ;
  }
  for (let c=0; c < board.length; c++) {
    if (board[r][c] != 'Q') {
      board[r][c] = 'Q'
      if (check(r-1, c)) {
        dfs(r+1)
      }
      board[r][c] = '.'
    }
  }
}
var check = function(r,c) {
  let slash = c+1; let backSlash = c-1

  while (-1 < r) {
    if (board[r][c] == 'Q') {
      return false
    } else if (slash < board.length && board[r][slash] == 'Q') {
      return false
    } else if (-1 < backSlash && board[r][backSlash] == 'Q'){
      return false
    }
    r -= 1; slash += 1; backSlash -= 1
  }
  return true
}

console.log(solveNQueens(5))